Stats:Improved parameter handling by Benini distribution

supreet11agrawal opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #16107 

#### Brief description of what is fixed or changed
Benini allows only positive shape and scale parameters. After this fix, non-positive parameters will result in an error.
Also, test cases for checking that valid parameter values are being given added.

#### Other comments
Total 8 possibilities for signs of alpha, beta and sigma. The tests include 5 out of them.
1) all positive
2) alpha negative
3) beta negative
4) alpha and beta negative
5) sigma negative
The error generated looks like `Shape parameters must be positive.` Should it explicitly say `alpha is negative`. I have not written this because the end user may change the names of parameters and get confused.

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* stats
  * Benini distribution now takes positive parameter values
<!-- END RELEASE NOTES -->

supreet11agrawal committed:
Improvised working of Benini distribution

Benini allows only positive parameters. After this change,
benini distribution won't run on non-positive parameters.
Test cases added for checking validity of parameters

supreet11agrawal committed:
Revert "Improvised working of Benini distribution"

This reverts commit 0196001435695c869e8f35fb466a542117411a6a.

supreet11agrawal committed:
Improvised parameter handling by Benini distribution

Benini allows only positive shape and scale parameters after this change.
Test cases also added to check that non positive parameter values should
give error.

jksuom commented on sympy/stats/crv_types.py:
```diff
@@ -232,6 +232,17 @@ def Arcsin(name, a=0, b=1):
     _argnames = ('alpha', 'beta', 'sigma')
 
+    def __new__(name, *args):
+        alpha = args[BeniniDistribution._argnames.index('alpha')]
```
Checking that there are exactly 3 arguments should be the first thing to do. Then it would suffice to write

    alpha, beta, sigma = args

supreet11agrawal replied:
Ok, I'll do it. And also please tell me should I change some of the things which I mentioned above under **Other Comments**

supreet11agrawal committed:
stats: Minor changes in parameter handling

Changed the condition from checking if not parameter.is_positive to
if not parameter > 0. This is to prevent the AttributeError: 'int' object
has no attribute 'is_positive'

supreet11agrawal commented:
@Upabjojr @jksuom Can you please review? There are some doubts which I wrote under **Other Comments**. Should I make those changes as well?

jksuom commented on sympy/stats/crv_types.py:
```diff
@@ -232,6 +232,15 @@ def Arcsin(name, a=0, b=1):
+        if not alpha > 0 or not beta > 0:
+            raise ValueError("Shape parameters must be positive.")
+        if not sigma > 0:
+            raise ValueError("Scale paramter must be positive.")
```
paramter -> parameter

supreet11agrawal committed:
Implemented check method for parametes

In earlier commit, the parameters were being checked by __new__ method.
Added a staticmethod check which does the same thing.

supreet11agrawal committed:
Removed spaces between test cases

oscargus commented on sympy/stats/crv_types.py:
```diff
@@ -232,6 +232,12 @@ def Arcsin(name, a=0, b=1):
+    def check(alpha, beta, sigma):
+        _value_check(alpha > 0, "Shape parameter Alpha must be positive.")
+        _value_check(beta > 0, "Shape parameter Beta must be positive.")
+        _value_check(sigma >0, "Scale parameter Sigma must be positive.")
```
space after >

supreet11agrawal replied:
Yup, will do that

smichr commented:
> 5\. Should it explicitly say `alpha `

I think what you have now is ok -- "shape...alpha...."

smichr commented:
> Total 8 possibilities for signs of alpha, beta and sigma. The tests include 5 out of them.

There are lots of ways to get something wrong. When the code is clearly written such that all parameters are positive, then 1 test showing that they don't trigger and another showing that they do trigger should be enough. Since there are 3 lines to trigger then 4 tests should be fine....which you get with 1, 2, 3 and 5. Test 4 will fail because of the first test (which was tested in 2) so it is redundant.

smichr commented on sympy/stats/tests/test_continuous_rv.py:
```diff
@@ -188,12 +188,25 @@ def test_arcsin():
+    assert density(X)(x) == ((alpha/x + 2*beta*log(x/sigma)/x)
+                          *exp(-alpha*log(x/sigma) - beta*log(x/sigma)**2))
+
+    alpha = Symbol("alpha", positive=False)
```
For me, something like this is more transparently correct:
```python
from sympy.utilities.iterables import rotate_left
npos = Symbol('np', positive=False)
args = [alpha, beta, npos]
assert [i.is_positive for i in args] == [1, 1, 0]
for _ in range(3):
    args = rotate_left(args, 1)
    raises(ValueError, lambda: Benini('x', *args)
```
YMMV

supreet11agrawal replied:
I think a newcomer might have problem understanding this. Still, if it is better then I'll do it.
Tell me if I should change it. I'll make the commit after that

supreet11agrawal commented:
>Test 4 will fail because of the first test (which was tested in 2) so it is redundant.
@smichr  I kept this test just because someone might make conditional statements in a future commit that can have such an exception. For eg `not False and not False`. It doesn't have much use though. I'll remove it if that makes it better.

supreet11agrawal committed:
Added a space after > in _value_check on line no. 239

jksuom commented:
Thanks, the tests could be organized differently but I think this is good to go.

Upabjojr commented:
Thanks for your contribution.
